compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -687,6 +687,24 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         &mut |err| {
                             if let Some(expected_ty) = expected.only_has_type(self) {
                                 self.consider_hint_about_removing_semicolon(blk, expected_ty, err);
+                                if expected_ty == self.tcx.types.bool {
+                                    // If this is caused by a missing `let` in a `while let`,
+                                    // silence this redundant error, as we already emit E0070.
+                                    let parent = self.tcx.hir().get_parent_node(blk.hir_id);
```

Comment:
this code seems really fragile :thinking: 

can you instead check the parent which should be a `Match` expr with `MatchSource::WhileDesugar`? at least that's what i expect

Reply:
We can't do this because this is *not* a `WhileDesugar`. It *would* be if we had an `while let`, but we instead have an `while`.

Reply:
An alternative would be to keep global state with all of the HirIds for which we should silence these errors 🤔